index/segment/mem/segment.go

```diff
@@ -0,0 +1,258 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package mem
+
+import (
+	"errors"
+	"regexp"
+	"sync"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index"
+	sgmt "github.com/m3db/m3ninx/index/segment"
+	"github.com/m3db/m3ninx/postings"
+)
+
+var (
+	errSegmentClosed     = errors.New("segment is closed")
+	errSegmentSealed     = errors.New("segment has been sealed")
+	errUnknownPostingsID = errors.New("unknown postings ID specified")
+)
+
+type segment struct {
+	opts   Options
+	offset int
+
+	wg sync.WaitGroup
+
+	// Internal state of the segment. The allowed transitions are:
+	//   - Open -> Sealed -> Closed
+	//   - Open -> Closed
+	state struct {
+		sync.RWMutex
+
+		sealed bool
+		closed bool
+	}
+
+	// Mapping of postings list ID to document.
+	docs struct {
+		sync.RWMutex
+
+		data []doc.Document
+	}
+
+	// Current writer and reader IDs. Writers increment the writer ID for each new
+	// document and only increment the reader ID after the document has been fully
+	// indexed by the segment. Readers do not need to acquire a lock.
+	ids struct {
```

Comment:
why have a lock at all if they're atomics?

Reply:
If we had only a single writer then the lock would be unnecessary since only that writer would be responsible for incrementing the IDs. With multiple writers the lock is used to enforce that only a single writer can increment both. I'll probably come back to this at some point. The main difficulty that needs to be solved is that with multiple writers we don't want to set the ID to a given value if documents with lower IDs haven't finished being indexed yet.

Reply:
OK, then how about not using atomic.ID for them? It's confusing to see an atomic guarded by a lock.

Reply:
Talked offline, will flesh out the comment to explain the semantics more.